### PR TITLE
feat: Add audit log object

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This pull request implements the audit log object

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [ ] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

**Comment by mAxYoLo01** [comment]
I feel like your file should be named `audit_log.py` so that Python gods are happy with the naming.

**Comment by sillyElvy** [comment]
> I feel like your file should be named `audit_log.py` so that Python gods are happy with the naming.

I named it as I named the scheduled event http file before xD

File: interactions/api/models/audit_log.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    :ivar int INVITE_DELETE: 42 - Server invite was deleted
+    :ivar int WEBHOOK_CREATE: 50 - Webhook was created
+    :ivar int WEBHOOK_UPDATE: 51 - Webhook properties or channel were updated
+    :ivar int WEBHOOK_DELETE: 52 - Webhook was deleted
+    :ivar int EMOJI_CREATE: 60 - Emoji was created
+    :ivar int EMOJI_UPDATE: 61 - Emoji name was updated
+    :ivar int EMOJI_DELETE: 62 - Emoji was deleted
+    :ivar int MESSAGE_DELETE: 72 - Single message was deleted
+    :ivar int MESSAGE_BULK_DELETE: 73 - Multiple messages were deleted
+    :ivar int MESSAGE_PIN: 74 - Message was pinned to a channel
```
Comment (Toricane) [inline interactions/api/models/audit_log.py:60]:
```suggestion
    :ivar int MESSAGE_PIN: 74 - Message was pinned in a channel
```
You don't pin a message to a channel. You pin a message that is in a channel.
Resolve if you just copied from discord docs

File: interactions/api/models/audit_log.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    reason: Optional[str] = field(default=None)
+
+
+@define()
+class AuditLogs(DictSerializerMixin):
+    """
+    A class object representing the audit logs of a guild.
+
+    :ivar List[AuditLogEntry] audit_log_entries: List of audit log entries, sorted from most to least recent.
+    :ivar List[AutoModerationRule] auto_moderation_rules:
```
Comment (Toricane) [inline interactions/api/models/audit_log.py]: Add desc

File: interactions/api/models/guild.py [hunk 3]
```diff
@@ -2271,6 +2272,160 @@ async def modify_auto_moderation_rule(
...
+            limit=limit,
+            before=_before,
+            user_id=_user_id,
+            action_type=_action_type,
+        )
+        return AuditLogs(**res)
+
+    async def get_latest_audit_log_action(
+        self,
+        of: Union[
```
Comment (Toricane) [inline interactions/api/models/guild.py:2316]: If it could be a tuple, why not *args?

File: interactions/api/models/guild.py [hunk 4]
```diff
@@ -2271,6 +2272,160 @@ async def modify_auto_moderation_rule(
...
+        self,
+        user_id: Optional[Union[User, int, Snowflake]] = MISSING,
+        action_type: Optional[Union[int, AuditLogEvents]] = MISSING,
+    ) -> AuditLogs:
+        """
+        Gets the full audit log of the guild.
+
+        :param user_id?: User ID snowflake. filter the log for actions made by a user.
+        :type user_id?: Optional[Union[User, int, Snowflake]]
+        :param action_type?: The Type of the audit log action.
```
Comment (Toricane) [inline interactions/api/models/guild.py]:
```suggestion
        :param action_type?: The type of the audit log action.
```

**Review by Toricane — CHANGES_REQUESTED** [review]
Some suggestions

File: interactions/api/models/guild.py (hunk 3 above)
Comment (sillyElvy) [reply interactions/api/models/guild.py:2316]: It can only be a Tuple of `(User, type)`, similar to members cache

Comment (Toricane) [reply interactions/api/models/guild.py:2316]: Why User? Why not Member?

Comment (sillyElvy) [reply interactions/api/models/guild.py:2316]: Because the param is named user_id and I just went with that, member works too tho, should I add it?

**Comment by sillyElvy** [comment]
> Some suggestions 

Done

File: interactions/api/models/guild.py (hunk 3 above)
Comment (Toricane) [reply interactions/api/models/guild.py:2316]: Sure

Comment (sillyElvy) [reply interactions/api/models/guild.py:2316]: Already done

File: interactions/api/models/guild.py [hunk 5]
```diff
@@ -2271,6 +2272,161 @@ async def modify_auto_moderation_rule(
...
+                guild_id=int(self.id), user_id=_user, action_type=of[1]
+            )
+
+        elif isinstance(of, AuditLogEvents) or isinstance(of, int) and len(str(of)) <= 3:
+            res = await self._client.get_guild_auditlog(
+                guild_id=int(self.id), limit=1, action_type=of
+            )
+
+        else:
+            if isinstance(of, User):
```
Comment (Toricane) [inline interactions/api/models/guild.py]:
```suggestion
            if isinstance(of, (Member, User}):
```

File: interactions/api/models/guild.py [hunk 6]
```diff
@@ -2271,6 +2272,161 @@ async def modify_auto_moderation_rule(
...
+
+        if isinstance(of, tuple):
+            if len(of) != 2 or len(str(of[1])) > 3:
+                raise LibraryException(
+                    12,
+                    "You specified invalid arguments in the tuple. Make sure the first argument"
+                    "is the user ID and the second is the action type!",
+                )
+
+            _user = of[0].id if isinstance(of[0], User) else of[0]
```
Comment (Toricane) [inline interactions/api/models/guild.py]:
```suggestion
            _user = of[0].id if isinstance(of[0], (Member, User)) else of[0]
```

**Review by Toricane — CHANGES_REQUESTED** [review]
No, it was not done

**Comment by sillyElvy** [comment]
> No, it was not done

Forgot about those, thanks 😅

**Review by Toricane — APPROVED** [review]
LGTM!

**Pull request merged**